controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -451,38 +571,130 @@ func (pp *portPublisher) updateEndpoints(endpoints *corev1.Endpoints) {
+				listener.Remove(remove)
+				for addrID := range remove.Addresses {
+					delete(pp.addresses.Addresses, addrID)
+				}
+			}
+			if len(add.Addresses) > 0 {
+				listener.Add(add)
+				for addrID, addr := range add.Addresses {
+					pp.addresses.Addresses[addrID] = addr
+				}
+			}
+		}
+		if len(pp.addresses.Addresses) == 0 && len(add.Addresses) > 0 {
+			pp.addresses = newAddressSet
+		}
+	}
 }
 
-func metricLabels(endpoints *corev1.Endpoints) map[string]string {
-	labels := map[string]string{service: endpoints.Name, namespace: endpoints.Namespace}
+func metricLabels(resource interface{}) map[string]string {
+	var serviceName, ns string
+	var resLabels, resAnnotations map[string]string
+	switch res := resource.(type) {
+	case *corev1.Endpoints:
+		{
+			serviceName, ns = res.Name, res.Namespace
+			resLabels, resAnnotations = res.Labels, res.Annotations
+		}
+	case *dv1beta1.EndpointSlice:
+		{
+			serviceName, ns = res.Labels[dv1beta1.LabelServiceName], res.Namespace
+			resLabels, resAnnotations = res.Labels, res.Annotations
+		}
+	}
+
+	labels := map[string]string{service: serviceName, namespace: ns}
 
-	gateway, hasRemoteGateway := endpoints.Labels[consts.RemoteGatewayNameLabel]
-	gatewayNs, hasRemoteGatwayNs := endpoints.Labels[consts.RemoteGatewayNsLabel]
-	remoteClusterName, hasRemoteClusterName := endpoints.Labels[consts.RemoteClusterNameLabel]
-	serviceFqn, hasServiceFqn := endpoints.Annotations[consts.RemoteServiceFqName]
+	gateway, hasRemoteGateway := resLabels[consts.RemoteGatewayNameLabel]
+	gatewayNs, hasRemoteGatwayNs := resLabels[consts.RemoteGatewayNsLabel]
+	remoteClusterName, hasRemoteClusterName := resLabels[consts.RemoteClusterNameLabel]
+	serviceFqn, hasServiceFqn := resAnnotations[consts.RemoteServiceFqName]
 
 	if hasRemoteGateway && hasRemoteGatwayNs && hasRemoteClusterName && hasServiceFqn {
 		// this means we are looking at Endpoints created for the purpose of mirroring
 		// an out of cluster service.
-		labels[targetGatewayNamespace] = gatewayNs
-		labels[targetGateway] = gateway
-		labels[targetCluster] = remoteClusterName
+		labels[remoteGatewayNamespace] = gatewayNs
+		labels[remoteGateway] = gateway
+		labels[remoteCluster] = remoteClusterName
 
 		fqParts := strings.Split(serviceFqn, ".")
 		if len(fqParts) >= 2 {
-			labels[targetService] = fqParts[0]
-			labels[targetServiceNamespace] = fqParts[1]
+			labels[remoteService] = fqParts[0]
+			labels[remoteServiceNamespace] = fqParts[1]
 		}
 	}
 	return labels
 }
 
+func (pp *portPublisher) endpointSliceToAddresses(es *dv1beta1.EndpointSlice) AddressSet {
+
+	addresses := make(map[ID]Address)
+	resolvedPort := pp.resolveESTargetPort(es.Ports)
+	for _, endpoint := range es.Endpoints {
+		if endpoint.Hostname != nil {
+			if pp.hostname != "" && pp.hostname != *endpoint.Hostname {
+				continue
+			}
+		}
+		if !*endpoint.Conditions.Ready {
+			continue
+		}
+
+		if endpoint.TargetRef == nil {
+			serviceName, err := getEndpointSliceSVCName(es)
```

Comment:
We can move that out of the loop is guess. No point of doing it for every endpoint as it will not change.

Reply:
That actually makes a lot of sense, changed! Thank you!